scripts/apps/authoring/authoring/authoring-custom-field.tsx

```diff
@@ -0,0 +1,47 @@
+import React from 'react';
+import {get} from 'lodash';
+
+import {getFields} from 'apps/fields';
+import {IArticle} from 'superdesk-interfaces/Article';
+import {IVocabulary} from 'superdesk-interfaces/Vocabulary';
+
+interface IProps {
+    item: IArticle;
+    field: IVocabulary;
+    onChange: (item: Partial<IArticle>, time: number) => any;
+    readonly: boolean;
+}
+
+export class AuthoringCustomField extends React.PureComponent<IProps> {
+    setValue(value) {
+        const extra = Object.assign({}, this.props.item.extra);
+
+        extra[this.props.field._id] = value || null;
+        this.props.item.extra = extra;
```

Comment:
it's problematic for storage if we add new fields dynamically, also to avoid users overriding some metadata which we need.

Reply:
but anyhow that value must be stored somewhere, and imo the component handling it is the place to do it, where else?

Reply:
> it's problematic for storage if we add new fields dynamically

Dynamic fields is a problem with pyeve, not anything else, right? What about not adding fields dynamically? Are there situations where it must be added dynamically?

Regarding unintentional overwriting of fields, what if an extension or two client repos want to add the same field to `extra`? It's not that unrealistic as we see with assignments when 2 clients have different implementations. It would be pretty easy to solve by requiring client instances/extensions to statically declare which fields they want to access and we could then verify on build that there are no conflicts.

> but anyhow that value must be stored somewhere, and imo the component handling it is the place to do it, where else?

I didn't get the question. Could you rephrase it? It's not a problem that you set the fields in this component, I was only commenting on `extra` field and not mutating the object,

Reply:
it will be added dynamically, we have UI for users to add new fields, not sure if we can avoid that

Reply:
You mean avoid unintentional overwriting of fields? Sure we can. We can keep a list of static fields in `package.json` or other file and validate that between client repositories/extensions and we can have dynamic fields in the DB which we can check against static fields on application start.

Or we can have one namespace for static and one for dynamic fields which was probably the idea behind `extra`, but it doesn't work unless `extra` reserved for dynamic fields only. Is that the case now?

Reply:
yes `extra` is only for dynamic fields

Reply:
Let's keep it then. Can you add the comment in the interface that explains it?

Reply:
We sidetracked into talking about the `extra` field and forgot about mutation. Can you update the `setValue` function so it doesn't mutate props? We shouldn't mutate props of React components in general. We might have to in some dirty angular integration code, but I think you should be able to easily avoid it here.